File: Data Structures & Algorithms/trapping-rain-water/submission-0.java
class Solution {
    public int trap(int[] height) {
        int[] maxLeft = new int[height.length];
        int[] maxRight = new int[height.length];

        maxLeft[0] = 0;
        for(int i=1; i<height.length; i++){
            maxLeft[i] = Math.max(maxLeft[i-1],height[i-1]);
        }
        maxRight[height.length-1]=0;
        for(int j=height.length-2; j>=0; j--){
            maxRight[j] = Math.max(maxRight[j+1],height[j+1]);
        }
        int resultArea = 0;
        for(int i=1; i<height.length; i++){
            int area = Math.min(maxLeft[i],maxRight[i]) - height[i];
            if(area>1){
                resultArea += area;
            }
        }
        return resultArea;
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-0.ts
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    maxAreaOfIsland(grid: number[][]): number {
        let maxAreaOfIsland = 0;

        const ROWS = grid.length;
        const COLS = grid[0].length;

        const directions = [
            [1, 0],
            [-1,0],
            [0,1],
            [0, -1]
        ]

        function dfs(r: number, c: number): number {
            if(r < 0 || c < 0 || r >= ROWS || c >= COLS || grid[r][c] === 0) {
                return 0;
            }
            grid[r][c] = 0;
            let area = 1;
            for(let [dr,dc] of directions) {
                area += dfs(r + dr, c + dc);
            }
            return area;
        }

        for(let r=0; r<ROWS; r++) {
            for(let c=0; c<COLS; c++) {
                if(grid[r][c] === 1) {
                    maxAreaOfIsland = Math.max(maxAreaOfIsland, dfs(r, c));
                }
            }
        }

        return maxAreaOfIsland;
    }
}
